File: asr/android/sdk-dingqiao/src/main/java/com/amphion/dingqiao/diarization/SpeakerDiarizationFinishBarrier.kt
package com.amphion.dingqiao.diarization

import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

internal data class DiarizationFinishInput<T>(val degraded: Boolean, val value: T)
internal data class DiarizationFinishOutput<A, S>(val asr: A, val speaker: S?, val degraded: Boolean)

/** Coordinates the ASR tail and diarization tail without blocking finish(). */
internal class SpeakerDiarizationFinishBarrier<A : Any, S : Any>(
    private val timeoutMs: Long,
    private val scheduler: ScheduledExecutorService,
    private val onReady: (DiarizationFinishOutput<A, S>) -> Unit,
    private val timeoutAsrFallback: (() -> A)? = null,
) {
    private var started = false
    private var completed = false
    private var asrReady = false
    private var speakerReady = false
    private var degraded = false
    private var asrValue: A? = null
    private var speakerValue: S? = null
    private var timeout: ScheduledFuture<*>? = null

    init { require(timeoutMs > 0) }

    @Synchronized
    fun begin() {
        if (started || completed) return
        started = true
        timeout = scheduler.schedule({
            synchronized(this) {
                if (completed) return@synchronized
                speakerReady = true
                degraded = true
                if (!asrReady) timeoutAsrFallback?.let {
                    asrValue = it()
                    asrReady = true
                }
                tryCompleteLocked()
            }
        }, timeoutMs, TimeUnit.MILLISECONDS)
    }

    @Synchronized
    fun resolveAsr(value: A) {
        if (completed || asrReady) return
        asrReady = true
        asrValue = value
        tryCompleteLocked()
    }

    @Synchronized
    fun resolveSpeaker(result: DiarizationFinishInput<S>) {
        if (completed || speakerReady) return
        speakerReady = true
        speakerValue = result.value
        degraded = result.degraded
        tryCompleteLocked()
    }

    @Synchronized
    fun cancel() {
        completed = true
        timeout?.cancel(false)
        timeout = null
    }

    private fun tryCompleteLocked() {
        if (completed || !asrReady || !speakerReady) return
        completed = true
        timeout?.cancel(false)
        timeout = null
        val output = DiarizationFinishOutput(checkNotNull(asrValue), speakerValue, degraded)
        onReady(output)
    }
}

File: asr/harmony/sdk-dingqiao/src/main/ets/com/amphion/dingqiao/diarization/SpeakerDiarizationFinishBarrier.ts
export interface SpeakerDiarizationFinishInput<T> {
  degraded: boolean;
  value: T;
}

export interface SpeakerDiarizationFinishOutput<A, S> {
  asr: A;
  speaker: S | undefined;
  degraded: boolean;
}

/** Coordinates the ASR tail and streaming diarization tail without blocking finish(). */
export class SpeakerDiarizationFinishBarrier<A, S> {
  private started: boolean = false;
  private completed: boolean = false;
  private asrReady: boolean = false;
  private speakerReady: boolean = false;
  private degraded: boolean = false;
  private asrValue?: A;
  private speakerValue?: S;
  private timer?: ReturnType<typeof setTimeout>;
  private readonly timeoutMs: number;
  private readonly onReady: (result: SpeakerDiarizationFinishOutput<A, S>) => void;

  constructor(
    timeoutMs: number,
    onReady: (result: SpeakerDiarizationFinishOutput<A, S>) => void,
  ) {
    if (timeoutMs <= 0) {
      throw new Error('timeoutMs must be positive');
    }
    this.timeoutMs = timeoutMs;
    this.onReady = onReady;
  }

  begin(): void {
    if (this.started || this.completed) {
      return;
    }
    this.started = true;
    this.startTimeoutIfReady();
  }

  private startTimeoutIfReady(): void {
    // Diarization also waits for the real ASR tail. Its timeout must not include ASR backlog.
    if (!this.started || !this.asrReady || this.speakerReady || this.completed ||
      this.timer !== undefined) return;
    this.timer = setTimeout(() => {
      if (this.completed || this.speakerReady) {
        return;
      }
      this.speakerReady = true;
      this.degraded = true;
      // Only diarization may degrade on timeout. ASR must drain all accepted audio.
      this.tryComplete();
    }, this.timeoutMs);
  }

  resolveAsr(value: A): void {
    if (this.completed || this.asrReady) {
      return;
    }
    this.asrReady = true;
    this.asrValue = value;
    this.startTimeoutIfReady();
    this.tryComplete();
  }

  resolveSpeaker(result: SpeakerDiarizationFinishInput<S>): void {
    if (this.completed || this.speakerReady) {
      return;
    }
    this.speakerReady = true;
    this.speakerValue = result.value;
    this.degraded = result.degraded;
    this.tryComplete();
  }

  private tryComplete(): void {
    if (this.completed || !this.asrReady || !this.speakerReady) {
      return;
    }
    this.completed = true;
    if (this.timer !== undefined) {
      clearTimeout(this.timer);
      this.timer = undefined;
    }
    this.onReady({
      asr: this.asrValue as A,
      speaker: this.speakerValue,
      degraded: this.degraded,
    });
  }
}
